File: app/Farert.android/app/src/main/java/org/sutezo/farert/Routefolder.kt
package org.sutezo.farert

import android.content.Context
import org.sutezo.alps.*

//
//  Routefolder.swift
//  Farert
//
//  Created by TAKEDA, Noriyuki on 2017/04/25.
//  Copyright © 2017年 TAKEDA, Noriyuki. All rights reserved.
//



class Routefolder {

    data class Folder(val routeList: RouteList, var aggregateType: Aggregate, var fare: Int? = null, var salesKm: Int? = null)

    //private static let onlyObj = Routefolder()
    
    enum class Aggregate {
        NORMAL,   // 普通運賃
        CHILD,     // 小児運賃
        ROUNDTRIP, // 往復
        STOCK,     // 株割
        STOCKW,    // 株割り4割
        ACADEMIC,   // 学割
        ACADEMIC_ROUNDTRIP,   // 学割往復
        NULLFARE,   // 無効
    }

    private val aggregate_label = arrayListOf("普通運賃","小児運賃","往復運賃","株割運賃","株割x2運賃","学割運賃","学割往復","無効")

    private var _routeList : MutableList<Folder> = arrayListOf()
    private var _totalFare : Int = 0
    private var _totalSalesKm : Int = 0

    private fun MutableList<Folder>.swap(index_a : Int, index_b: Int) {
        val tmp = this[index_a]
        this[index_a] = this[index_b]
        this[index_b] = tmp
    }

    fun totalFare() : String {
        return if (_routeList.count() <= 0) {
            "¥---,---"
        } else {
            "¥${fareNumStr(_totalFare)}"
        }
    }

    fun totalSalesKm() : String {
        return if (_routeList.count() <= 0) {
            "---,--- km"
        } else {
            "${kmNumStr(_totalSalesKm)} km"
        }
    }

    fun add(context: Context, item : RouteList) : Boolean {
        if (MAX_HOLDER <= _routeList.count()) {
            return false
        }
        _routeList.add(Folder(item, Aggregate.NORMAL))
        calc()
        save(context)
        return true
    }

    fun swap(context: Context, from: Int, to: Int) : Boolean {
        if (from < 0 || to < 0 || _routeList.count() < from || _routeList.count() < to) {
            return false
        }
        _routeList.swap(from, to)
        calc()
        save(context)
        return true
    }

    //fun route(index : Int) -> cRoute {
    //    if index < _routeList.count {
    //        return cRoute(_routeList: _routeList[index]._routeList)
    //    } else {
    //        assert(false, "Index failure...")
    //        return cRoute()
    //   }
    //}

    fun routeItem(index : Int) : RouteList {
        return if (index < _routeList.count()) {
            _routeList[index].routeList
        } else {
            assert(false) // , "Index failure...")
            RouteList()
        }
    }

    fun routeItemFareKm(index : Int) : Pair<Int, Int> {
        return if (index < _routeList.count()) {
            if ((_routeList[index].fare != null) && (_routeList[index].salesKm != null)) {
                Pair(_routeList[index].fare!!, _routeList[index].salesKm!!)
            } else {
                val fare = calcFare(_routeList[index])
                fare
            }
        } else {
            assert(false)//, "Index failure...")
            Pair(0, 0)
        }
    }

    fun count() : Int {
        return _routeList.count()
    }

    fun aggregateType(index : Int) : Aggregate {
        return if (index < _routeList.count()) {
            _routeList[index].aggregateType
        } else {
            Aggregate.NULLFARE
        }
    }

    fun setAggregateType(context: Context, index : Int, aggr : Aggregate) {
        if ((index < _routeList.count()) && (_routeList[index].aggregateType != aggr)) {
            _routeList[index].aggregateType = aggr
            // do re-calc fare
            _routeList[index].fare = null
            _routeList[index].salesKm = null

            calc()
            save(context)
        }
    }

    fun remove(context: Context, index : Int, no_write: Boolean = false) {
        _routeList.removeAt(index)
        calc()
        if (!no_write) {
            save(context)
        }
    }

    fun removeAll(context: Context) {
        _totalSalesKm = 0
        _totalFare = 0
        _routeList.removeAll { true }
        calc()
        save(context)
    }

    private fun calc() {
        var fare_sum = 0
        var salesKm_sum = 0
        for (route in this._routeList) {
            if ((route.fare != null) && (route.salesKm != null)) {
                fare_sum += route.fare!!
                salesKm_sum += route.salesKm!!
            } else {
                val result = calcFare(route)
                fare_sum += result.first
                salesKm_sum += result.second
                route.fare = result.first
                route.salesKm = result.second
            }
        }
        _totalFare = fare_sum
        _totalSalesKm = salesKm_sum
    }

    fun save(context: Context) {
        val folders : MutableList<String> = mutableListOf()
        val dbver = DatabaseOpenHelper.DATABASE_VERSION
        val strDbVer = "DBVer|$dbver\n"
        folders.add(strDbVer)
        for (one in _routeList) {
            val rs = if ((one.fare != null) && (one.salesKm != null)) {
                one.aggregateType.ordinal.toString() + "|" +
                        one.routeList.route_script() + "|" +
                        one.fare.toString() + "|" + one.salesKm.toString() + "\n"
            } else {
                one.aggregateType.ordinal.toString() + "|" +
                        one.routeList.route_script() + "\n"
            }
            folders.add(rs)
        }
        saveParam(context, "folder", folders.toList())
    }

    fun load(context: Context, doCalc: Boolean? = false) {
        val folders = readParams(context, "folder")
        var isDbChanged : Boolean = doCalc ?: false
        _routeList.removeAll { true }

        for (o in folders) {
            val item = o.split("|")
            val nCol = item.count()
            if ((nCol != 2) && (nCol != 4)) {
                continue
            }
            if ((nCol == 2) && (0 == _routeList.count())) {
                if (item[0] == "DBVer") {
                    val dbver = DatabaseOpenHelper.DATABASE_VERSION
                    val folderDbVer = item[1].trim { c -> c == '\n'}.toInt()
                    if (dbver != folderDbVer) {
                        isDbChanged = true
                    }
                    continue
                }
            }
            val ord : Int = try {
                item[0].toInt()
            } catch (e: NumberFormatException) {
                0
            }
            val agr: Aggregate = Aggregate.values()[ord]
            val script = item[1].trim { c -> c == '\n' }
            val route = Route()
            val rc = route.setup_route(script)
            if (rc < 0) {
                continue
            }
            var fare : Int? = null
            var salesKm : Int? = null
            if (!isDbChanged && (nCol == 4)) {
                fare = Integer.parseInt(item[2])
                salesKm = Integer.parseInt(item[3].trim { c -> c == '\n' })
            }
            val folder = Folder(route, agr, fare, salesKm)
            _routeList.add(folder)

            if (MAX_HOLDER <= _routeList.count()) {
                break
            }
        }
        calc()
    }

    fun calcFare(item : Folder) : Pair<Int, Int> {

        val cds = CalcRoute(item.routeList)
        cds.let {
            val fareInfo = it.calcFareInfo()
            val fare : Int

            when (item.aggregateType) {
                Aggregate.NORMAL -> {   // 普通運賃
                    fare = fareInfo.fare
                }
                Aggregate.CHILD -> {     // 小児運賃
                    fare = fareInfo.childFare
                }
                Aggregate.ROUNDTRIP -> { // 往復
                    fare = fareInfo.roundTripFareWithCompanyLine
                }
                Aggregate.STOCK -> {     // 株割
                    val stocks = fareInfo.fareForStockDiscounts ?: listOf()
                    fare =
                        if (1 <= stocks.count()) {
                            // title, normal_fare, 114applied_fare
                            if (fareInfo.isRule114Applied) stocks[0].third else stocks[0].second    // thirdは114適用
                        } else {
                            fareInfo.fare
                        }
                }
                Aggregate.STOCKW -> {    // 株割り4割
                    val stocks = fareInfo.fareForStockDiscounts ?: listOf()
                    fare =
                        if (2 <= stocks.count()) {
                            if (fareInfo.isRule114Applied) stocks[1].third else stocks[1].second
                        } else if (1 <= stocks.count()) {
                            if (fareInfo.isRule114Applied) stocks[0].third else stocks[0].second
                        } else {
                            fareInfo.fare
                        }
                }
                Aggregate.ACADEMIC -> {    // 学割
                    fare = if (fareInfo.isAcademicFare) {
                        fareInfo.academicFare
                    } else {
                        fareInfo.fare
                    }
                }
                Aggregate.ACADEMIC_ROUNDTRIP -> {    // 学割往復
                    fare = if (fareInfo.isAcademicFare) {
                        fareInfo.roundtripAcademicFare
                    } else {
                        fareInfo.roundTripFareWithCompanyLine
                    }
                }
                else -> {    // 無効
                    return Pair(0, 0)
                }
            }
            return Pair(fare, fareInfo.totalSalesKm)
        }
        return Pair(0, 0)
    }

    fun makeExportText() : String {
        val result = mutableListOf<String>()

        var fare_sum = 0
        var salesKm_sum = 0
        for (route in this._routeList) {
            val cols = mutableListOf<String>()
            var fare : Int
            var salesKm : Int
            if ((route.fare != null) && (route.salesKm != null)) {
                fare = route.fare!!
                salesKm = route.salesKm!!
            } else {
                val fare_salesKm: Pair<Int, Int> = this.calcFare(route)
                fare = fare_salesKm.first
                salesKm = fare_salesKm.second
            }
            fare_sum += fare
            salesKm_sum += salesKm

            cols.add(aggregate_label[route.aggregateType.ordinal]) // 運賃
            cols.add("¥${fareNumStr(fare)}")
            cols.add("営業キロ")
            cols.add("${kmNumStr(salesKm)}km")

            val s = route.routeList.route_script().replace(",", " ")
            cols.add(s)
            result.add(cols.joinToString(","))
        }
        val cols = mutableListOf<String>()
        cols.add("総額")
        cols.add("¥${fareNumStr(fare_sum)}")
        cols.add("総営業キロ")
        cols.add("${kmNumStr(salesKm_sum)}km")
        result.add(cols.joinToString(","))
        return result.joinToString("\n")
    }
}
    

File: app/Farert.ios/Farert/Routefolder.swift
//
//  Routefolder.swift
//  Farert
//
//  Created by TAKEDA, Noriyuki on 2017/04/25.
//  Copyright © 2017年 TAKEDA, Noriyuki. All rights reserved.
//

import Foundation

public enum Aggregate: Int {
    case NORMAL = 0   // 普通運賃
    case CHILD     // 小児運賃
    case ROUNDTRIP // 往復
    case STOCK     // 株割
    case STOCKW    // 株割り4割
    case ACADEMIC   // 学割
    case ACADEMIC_ROUNDTRIP   // 学割往復
    case NULLFARE   // 無効
}

public let aggregate_label = [
"普通運賃", "小児運賃", "往復運賃", "株割運賃", "株割x2運賃",
                    "学割運賃", "学割往復" , "無効"]

struct folder {
    var routeList : cRouteList
    var indexOfAggregate : Int
    var fare : Int?
    var salesKm : Int?
}

public class Routefolder {
    
    //private static let onlyObj = Routefolder()
    
    private var routeList : [folder] = []
    private var _totalFare : Int = 0
    private var _totalSalesKm : Int = 0

    var totalFare : String {
        get {
            if routeList.count <= 0 {
                return "¥---,---"
            } else {
                let formatter = NumberFormatter()
                formatter.numberStyle = NumberFormatter.Style.decimal
                formatter.groupingSeparator = ","
                formatter.groupingSize = 3
                return "¥\(formatter.string(from: NSNumber(value: _totalFare)) ?? "")"
                //return "¥\(String(describing: formatter.string(from: NSNumber(value: _totalFare))))"
            }
        }
    }
    
    var totalSalesKm : String {
        get {
            if routeList.count <= 0 {
                return "---,--- km"
            } else {
                return "\(String(describing: cRouteUtil.kmNumStr(_totalSalesKm)!)) km"
            }
        }
    }

    func add(item : cRouteList) -> Bool {
        if Int(MAX_HOLDER) <= routeList.count {
            return false
        }
        routeList.append(folder(routeList: item, indexOfAggregate: Aggregate.NORMAL.rawValue))
        calc()
        save()
        return true
    }
    
    //func route(index : Int) -> cRoute {
    //    if index < routeList.count {
    //        return cRoute(routeList: routeList[index].routeList)
    //    } else {
    //        assert(false, "Index failure...")
    //        return cRoute()
    //   }
    //}
    
    func routeItem(index : Int) -> cRouteList {
        if index < routeList.count {
            return routeList[index].routeList
        } else {
            assert(false, "Index failure...")
            return cRouteList()
        }
    }

    func routeItemFare(index : Int) -> (fare : String, salesKm : String) {
        if index < routeList.count {
            var fare : Int
            var salesKm : Int

            if ((routeList[index].fare != nil) && (routeList[index].salesKm != nil)) {
                fare = routeList[index].fare!
                salesKm = routeList[index].salesKm!
            } else {
                let fareSalesKm = self.calcFare(item: routeList[index])
                fare = fareSalesKm.fare
                salesKm = fareSalesKm.sales_km
            }
            let formatter = NumberFormatter()
            formatter.numberStyle = NumberFormatter.Style.decimal
            formatter.groupingSeparator = ","
            formatter.groupingSize = 3
            return ("¥\(formatter.string(from: NSNumber(value: fare)) ?? "")-",
                    "\(String(describing: cRouteUtil.kmNumStr(salesKm)!)) km")
        } else {
            assert(false, "Index failure...")
            return ("", "")
        }
    }
    
    func count() -> Int {
        return routeList.count
    }
 
    func aggregateType(index : Int) -> Int {
        if index < routeList.count {
            return routeList[index].indexOfAggregate
        } else {
            return 0
        }
    }

    func setAggregateType(index : Int, aggr : Int) -> Void {
        if ((index < routeList.count) && (routeList[index].indexOfAggregate != aggr)) {
            routeList[index].indexOfAggregate = aggr
            // do re-calc fare
            routeList[index].fare = nil
            routeList[index].salesKm = nil
            calc()
            save()
        }
    }
    
    func remove(index : Int) {
        routeList.remove(at: index)
        calc()
        save()
    }
    
    func removeAll() -> Void {
        _totalSalesKm = 0
        _totalFare = 0
        routeList.removeAll()
        calc()
        save()
    }
   
    /*  Exchange content index_a and index b
     *
     */
    func exchange(index_source : Int, index_destination : Int) {
        let obj = routeList[index_source]
        routeList.remove(at: index_source)
        routeList.insert(obj, at: index_destination)
        save()
    }
    

    func calc() {
        var fare_sum : Int = 0
        var salesKm_sum : Int = 0
        for (i, route) in self.routeList.enumerated() {
            if ((route.fare != nil) && (route.salesKm != nil)) {
                fare_sum += route.fare!
                salesKm_sum += route.salesKm!
            } else {
                let result = self.calcFare(item: route)
                fare_sum += result.fare
                salesKm_sum += result.sales_km

                self.routeList[i].fare = result.fare
                self.routeList[i].salesKm = result.sales_km
            }
        }
        self._totalFare = fare_sum
        self._totalSalesKm = salesKm_sum
    }
    
    private var filePath : String {
        get {
            let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0] as String
            return documentsPath + "/routeholder.txt"
        }
    }

    func save() {
        //print(filePath)
        if let os = OutputStream(toFileAtPath: filePath, append: false) {
            os.open()
            let dbver = cRouteUtil.getDatabaseId().rawValue
            let strDbVer = "DBVer|" + String(dbver) + "\n"
            os.write(strDbVer, maxLength: strDbVer.count)
            for one in routeList {
                let rs : String
                if ((one.fare != nil) && (one.salesKm != nil)) {
                    rs = String(one.indexOfAggregate) + "|" + one.routeList.routeScript()
                     + "|" + String(one.fare!) + "|" + String(one.salesKm!) + "\n"
                } else {
                    rs = String(one.indexOfAggregate) + "|" + one.routeList.routeScript() + "\n"
                }
                os.write(rs, maxLength: rs.lengthOfBytes(using: String.Encoding.utf8))
            }
            os.close()
        }
    }
    
    func load(_ doCalc : Bool? = nil) {
        do {
            self.routeList.removeAll()
            var num : Int = 0
            var isDbChanged : Bool = doCalc ?? false
            let text = try String(contentsOf: URL(fileURLWithPath: filePath), encoding: String.Encoding.utf8)
            text.enumerateLines { line, stop in
                if num < Int(MAX_ARCHIVE_ROUTE) {
                    let record : [String] = line.components(separatedBy: "|")
                    if 2 <= record.count {
                        if ((num == 0) && ("DBVer" == String(record[0]))) {
                            let dbVer = cRouteUtil.getDatabaseId().rawValue
                            if dbVer != Int(record[1]) {
                                isDbChanged = true
                            }
                        } else {
                            if let agr : Int = Int(record[0]) {
                                if let rt = cRoute() {
                                    let result = rt.setupRoute(record[1])
                                    if 0 <= result {
                                        if (!isDbChanged && (4 == record.count)) {
                                            let fare = Int(record[2])
                                            let salesKm = Int(record[3])
                                            self.routeList.append(
                                                folder(routeList: cRouteList(route: rt),
                                                       indexOfAggregate: agr,
                                                       fare: fare,
                                                       salesKm: salesKm))
                                        } else {
                                            self.routeList.append(folder(routeList: cRouteList(route: rt), indexOfAggregate: agr))
                                        }
                                        num += 1
                                    }
                                }
                            }
                        }
                    }
                }
            }
        } catch {
            // Failed to read file

            print("read fail!!!!!!!!!")
        }
        self.calc()
    }

    func calcFare(item : folder) -> (fare : Int, sales_km : Int) {

        if let cds = cCalcRoute(routeList: item.routeList) {
            if let fareInfo : FareInfo = cds.calcFare() {
                var fare : Int = 0

                switch item.indexOfAggregate {
                case Aggregate.NORMAL.rawValue:   // 普通運賃
                    fare = fareInfo.fare
                case Aggregate.CHILD.rawValue:     // 小児運賃
                    fare = fareInfo.childFare
                case Aggregate.ROUNDTRIP.rawValue: // 往復
                    fare = fareInfo.roundTripFareWithCompanyLine
                case Aggregate.STOCK.rawValue:     // 株割
                    if fareInfo.availCountForFareOfStockDiscount < 1 {
                        // N/A
                        fare = fareInfo.fare
                    } else {
                        if (fareInfo.isRule114Applied) {
                            //株割 有効 114条有効
                            fare = fareInfo.fare(forStockDiscount: 0+2)
                        } else {
                            //株割 有効 114条無効
                            fare = fareInfo.fare(forStockDiscount: 0)
                        }
                    }
                case Aggregate.STOCKW.rawValue:    // 株割り4割
                    if fareInfo.availCountForFareOfStockDiscount < 1 {
                        // N/A
                        fare = fareInfo.fare
                    } else if 1 < fareInfo.availCountForFareOfStockDiscount {
                        // 株割4
                        if (fareInfo.isRule114Applied) {
                            fare = fareInfo.fare(forStockDiscount: 1+2)
                        } else {
                            fare = fareInfo.fare(forStockDiscount: 1)
                        }
                    } else {
                        // 株割4は無効で株割2のみ有効
                        if (fareInfo.isRule114Applied) {
                            fare = fareInfo.fare(forStockDiscount: 0+2)
                        } else {
                            fare = fareInfo.fare(forStockDiscount: 0)
                        }
                    }
                case Aggregate.ACADEMIC.rawValue:    // 学割
                    if fareInfo.isAcademicFare {
                        fare = fareInfo.academicFare
                    } else {
                        fare = fareInfo.fare
                    }
                case Aggregate.ACADEMIC_ROUNDTRIP.rawValue:    // 学割往復
                    if fareInfo.isAcademicFare {
                        fare = fareInfo.roundtripAcademicFare
                    } else {
                        fare = fareInfo.roundTripFareWithCompanyLine
                    }
                default:    // 無効
                    return (0, 0)
                }
                return (fare, fareInfo.totalSalesKm)
            }
        }
        return (0, 0)
    }
    
    func conmaNumStr(_ num : Int) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = NumberFormatter.Style.decimal
        formatter.groupingSeparator = ","
        formatter.groupingSize = 3
        return "¥\(formatter.string(from: NSNumber(value: num)) ?? "")"
    }
    
    func makeExportText() -> String {
        var result : String = ""
        
        var fare_sum : Int = 0
        var salesKm_sum : Int = 0
        for folder_ in self.routeList {
            let fare : Int
            let salesKm : Int
            if ((folder_.salesKm != nil) && (folder_.fare != nil)) {
                fare = folder_.fare!
                salesKm = folder_.salesKm!
            } else {
                let fare_ = self.calcFare(item: folder_)
                fare = fare_.fare
                salesKm = fare_.sales_km
            }
            var cols = Array<String>()
            fare_sum += fare
            salesKm_sum += salesKm

            cols.append(aggregate_label[folder_.indexOfAggregate]) // 運賃
            cols.append(conmaNumStr(fare))
            cols.append("営業キロ")
            cols.append("\(String(describing: cRouteUtil.kmNumStr(salesKm_sum)!)) km")

            let s = folder_.routeList.routeScript()?.replacingOccurrences(of: ",", with: " ", options: .literal, range: nil) ?? ""
            cols.append(s)
            result += (cols.joined(separator: ",") + "\n")
        }
        var cols = Array<String>()
        cols.append("総額")
        cols.append(conmaNumStr(fare_sum))
        cols.append("総営業キロ")
        cols.append("\(String(describing: cRouteUtil.kmNumStr(salesKm_sum)!)) km")
        result += (cols.joined(separator: ",") + "\n")
        return result
    }

}
